File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    void dfs(int node, vector<vector<int>> &adjls, vector<int> &vis){
        vis[node] = 1;

        for(auto it : adjls[node]){
            if(!vis[it]){
                dfs(it, adjls, vis);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<vector<int>> adjls(n);

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(isConnected[i][j] == 1 && i != j){
                    adjls[i].push_back(j);
                    adjls[j].push_back(i);
                }
            }
        }

        int provinces = 0;
        vector<int> vis(n);

        for(int i = 0; i < n; i++){
            if(!vis[i]){
                dfs(i, adjls, vis);
                provinces++;
            }
        }

        return provinces;
    }
};

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def dfs(self, node, vis, adj):
        vis[node] = 1

        for it in adj[node]:
            if vis[it] == 0:
                self.dfs(it, vis, adj)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        vis = [0] * n
        adj = [[] for _ in range(n)]
        
        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1 and i != j:
                    adj[i].append(j)
                    adj[j].append(i)

        count = 0
        for i in range(n):
            if vis[i] == 0:
                self.dfs(i, vis, adj)
                count += 1
        
        return count